Fix file drag & drop crash for unsupported media types

ezxyz opened the pull request:
##  Summary
Fixes the drag and drop crash issue when an unsupported file type is dragged into the media pane (for example a .txt file). 

## Relevant Technical Choices

The issue was originally happening because "filterResources" in media pane was being called with 'null". I made changes so that:
1) Null should never be put inside the "media" context (via. uploadMedia).
2) Filter resources can handle null argument (just in case null is being passed into media context in some other way)
3) Made the "processor" function in mediaProvider, be skipped if there is a null (this results in a warning if there is a null value in the media file).

This meant I had to filter out unsupported filetypes from uploadMedia. This part was unit tested.

Fixes #1342

Commit:
Fixed issue with unsupported media type crashing media pane.

Commit:
Changed mediaPane.js

Commit:
Update useUploadMedia.js

Commit:
Added tests for skipping unsupported media uploads.

Commit:
Fixed eslint issues

Commit:
change "toEqual" to "toStrictEqual" as suggested by ESLint.

Commit:
Fixed more ESLint issues.

Commit:
Added header to new useUploadMedia.test.js (Linter issues)

spacedmonkey commented on assets/src/edit-story/app/media/mediaProvider.js:
```diff
@@ -126,15 +126,19 @@ function MediaProvider({ children }) {
           id
         ) {
-          await uploadVideoPoster(id, src);
+          uploadVideoPoster(id, src);
```
If we don't await this promise here, does it not run?

ezxyz replied:
I removed this one because uploadVideoPoster() was not async as far as I can tell. If you look above, the callback isn't an asynchronous function and hence doesn't return a promise. (It does create an asynchronous function inside the callback and then call it, but it doesn't await on the response of that promise).

Hence I thought putting await here wasn't doing anything.

ezxyz replied:
On second thought, I decided to change this back to the previous version (with an async await) since I'd prefer to keep this PR related to just the File drag and drop crash issue.

spacedmonkey commented on assets/src/edit-story/app/media/useUploadMedia/useUploadMedia.js:
```diff
@@ -62,6 +62,8 @@ function useUploadMedia({ media, pagingNum, mediaType, fetchMedia, setMedia }) {
             localResource: await getResourceFromLocalFile(file),
             file,
           }))
+        ).then((newLocalFiles) =>
```
Can you use await and then together like this?

ezxyz replied:
https://codesandbox.io/s/nostalgic-meitner-6g4e3?file=/src/index.js

I tried something similar here to show what happens.

Then only gets run after the promise gets fulfilled.  The promise will only get fulfilled after the getResourceFromLocalFile gets resolved.

See here: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/then

swissspidy replied:
It works, although it could just as well be:

```js
localFiles = await Promise.all(
  files.reverse().map(async (file) => ({
    localResource: await getResourceFromLocalFile(file),
    file,
  }))
);
localFiles = localFiles.filter((f) => f.localResource != null)
```

edit: fixed code sample as per Jon's comment

ezxyz replied:
So the actual code for that will be

`localFiles = localFiles.filter((...))` since filter actually returns the filtered array and doesn't change the actual array itself.

I personally prefered using "then" as it shows that its tied to the promise. However if you feel strongly about this, I'm happy to change it.

spacedmonkey replied:
I think it is confusing to have an await and a then. In my mind, it is one of the other. I don't think either way, just pick one.

ezxyz replied:
Chose the localFiles = localFiles.filter version.

Commit:
Reverted the async process back (in mediaProvider.js)

swissspidy commented on assets/src/edit-story/app/media/test/useUploadMedia.test.js:
```diff
@@ -0,0 +1,110 @@
+/*
```
Nit: No need for the `.test` suffix in the file name

dvoytenko replied:
That came from me. I'm ok to drop that pattern if others are feeling strongly.

ezxyz replied:
Done.

swissspidy commented on assets/src/edit-story/app/media/mediaProvider.js:
```diff
@@ -126,7 +126,11 @@ function MediaProvider({ children }) {
   const processor = useCallback(
-    ({ mimeType, posterId, id, src, local }) => {
+    (el) => {
+      if (!el) {
```
Alternatively do something like `return accumulatorPromise.then(() => el && processor(el));` below? And let `processor` assume it always gets passed a valid element

ezxyz replied:
SGTM. I've added that implementation.

Commit:
(mediaProvider.js) processer now assumes valid input is recieved.

spacedmonkey committed:
Merge branch 'master' into file-drag-drop-crash

dvoytenko commented on assets/src/edit-story/components/library/panes/media/mediaPane.js:
```diff
@@ -212,7 +212,11 @@ function MediaPane(props) {
   const filterResource = useCallback(
-    ({ mimeType, width, height }) => {
+    (resource) => {
+      if (!resource) {
```
This seems like hiding a problem. Why would we have `null` resources? I'd rather we fix at the source.

ezxyz replied:
Fixed it so we don't have null resource.

Commit:
Resolving PR comments.

1) Changing useUploadMidea.test.js > useUploadMedia.js
2) Removing catcing of null in filterresource since the problem is dealt with at the source.
3) Removing "then" from useUploadMedia promise.

Commit:
Merge branch 'file-drag-drop-crash' of https://github.com/google/web-stories-wp into file-drag-drop-crash

Commit:
Making localFiles throw error instead of silently failing so that snackbar is shown.

Commit:
using filter for localfiles

barklund commented on assets/src/edit-story/app/media/test/useUploadMedia.js:
```diff
@@ -0,0 +1,110 @@
+
+    expect(setMedia).toHaveBeenCalledTimes(1);
+
+    const sortedMediaArgs = setMedia.mock.calls[0][0].media.sort(sortBySrc);
```
With [`toIncludeSameMembers()` from `jest-extended`](https://github.com/jest-community/jest-extended#toincludesamemembersmembers), you could avoid all this sorting both here and above. It's cleaner, IMHO.

ezxyz replied:
Done.

Commit:
Added and used jest-extended for testing. (toIncludeSameMembers)

Commit:
Update jest.setup.js

spacedmonkey reviewed (approved):
LGTM
